From 11543b3b1db62d6d249f99f745e1e46bb2b1ed2c Mon Sep 17 00:00:00 2001
From: Jan Moeller <Jan.Moeller@governikus.de>
Date: Tue, 26 May 2026 14:51:42 +0200
Subject: a11y Android: Inform TalkBack of scrolled viewport

For "indexed" scroll element like lists, the to/from index and item count
properties of Android's AccessibilityEvent are set. Other scroll elements
use the scroll x/y and max scroll x/y properties to communicate their
viewport to TalkBack.

Task-number: QTBUG-139833
Pick-to: 6.12
Change-Id: I6a13f641b907f2a88eaf42b41b70abd2d6365a4e
Reviewed-by: Assam Boudjelthia <assam.boudjelthia@qt.io>
(cherry-picked from commit 68bc0af327415489c0d482094656a0c8ebadbe7d)
---
 .../qt/android/QtAccessibilityDelegate.java   |  3 +
 .../qt/android/QtNativeAccessibility.java     |  2 +
 .../android/androidjniaccessibility.cpp       | 84 ++++++++++++++++++-
 3 files changed, 88 insertions(+), 1 deletion(-)

diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
index b17b7ce77bb44309418f27cc3b3d4e6db3c54cdc..79626a38ef06cd3a5046140b5b6cb8aad0c2ebb5 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
@@ -373,6 +373,9 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
         String description = QtNativeAccessibility.descriptionForAccessibleObject(virtualViewId);
         event.setContentDescription(addLocaleSpan(virtualViewId, description));
 
+        if (eventType == AccessibilityEvent.TYPE_VIEW_SCROLLED)
+            QtNativeAccessibility.populateScrollEvent(virtualViewId, event);
+
         if (event.getText().isEmpty() && TextUtils.isEmpty(event.getContentDescription()))
             Log.w(TAG, "AccessibilityEvent with empty description");
 
diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
index 76c442a47513b7066b5ad26dc8e61ed6b64af8dd..c0d180ffe5f565dfd801132054d1d6a1491ae0e6 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
@@ -4,6 +4,7 @@
 package org.qtproject.qt.android;
 
 import android.graphics.Rect;
+import android.view.accessibility.AccessibilityEvent;
 import android.view.accessibility.AccessibilityNodeInfo;
 
 class QtNativeAccessibility
@@ -25,4 +26,5 @@ class QtNativeAccessibility
     static native boolean collapse(int objectId);
 
     static native boolean populateNode(int objectId, AccessibilityNodeInfo node);
+    static native boolean populateScrollEvent(int objectId, AccessibilityEvent event);
 }
diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index b3fd28b9023680c97ab5f2ca3cc6f8ecdc853289..f1998ac478bbae24ae6feb6c6874166335cf1bd6 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -45,6 +45,13 @@ namespace QtAndroidAccessibility
     static jmethodID m_setTextSelectionMethodID = 0;
     static jmethodID m_setRangeInfoMethodID = 0;
     static jmethodID m_setVisibleToUserMethodID = 0;
+    static jmethodID m_setMaxScrollXMethodID = 0;
+    static jmethodID m_setScrollXMethodID = 0;
+    static jmethodID m_setMaxScrollYMethodID = 0;
+    static jmethodID m_setScrollYMethodID = 0;
+    static jmethodID m_setItemCount = 0;
+    static jmethodID m_setFromIndex = 0;
+    static jmethodID m_setToIndex = 0;
 
     static int RANGE_TYPE_INT = 0;
     static int RANGE_TYPE_FLOAT = 0;
@@ -806,6 +813,8 @@ namespace QtAndroidAccessibility
                 info.actions.contains(QAccessibleActionInterface::increaseAction());
         const bool hasDecreaseAction =
                 info.actions.contains(QAccessibleActionInterface::decreaseAction());
+        const bool scrollableRole =
+                info.role == QAccessible::ScrollBar || info.role == QAccessible::List;
 
         if (info.hasTextSelection && m_setTextSelectionMethodID) {
             env->CallVoidMethod(node, m_setTextSelectionMethodID, info.selectionStart,
@@ -856,7 +865,8 @@ namespace QtAndroidAccessibility
         if (m_setHeadingMethodID)
             env->CallVoidMethod(node, m_setHeadingMethodID, info.role == QAccessible::Heading);
         env->CallVoidMethod(node, m_setVisibleToUserMethodID, !info.state.invisible);
-        env->CallVoidMethod(node, m_setScrollableMethodID, hasIncreaseAction || hasDecreaseAction);
+        env->CallVoidMethod(node, m_setScrollableMethodID,
+                            hasIncreaseAction || hasDecreaseAction || scrollableRole);
         env->CallVoidMethod(node, m_setClickableMethodID, hasClickableAction || info.role == QAccessible::Link);
 
         // Add ACTION_CLICK
@@ -891,6 +901,68 @@ namespace QtAndroidAccessibility
         return true;
     }
 
+    struct ScrollEventInfo
+    {
+        bool isValid = false;
+        bool isIndexed;
+        QSizeF contentSize;
+        QPointF position;
+        QSizeF viewportSize;
+    };
+
+    static ScrollEventInfo populateScrollEvent_helper(int objectId)
+    {
+        QAccessibleInterface *iface = interfaceFromId(objectId);
+        if (!iface || !iface->isValid())
+            return { };
+
+        auto *viewportIface = iface->viewportInterface();
+        if (!viewportIface)
+            return { };
+
+        ScrollEventInfo info;
+        info.isValid = true;
+        info.isIndexed = viewportIface->isIndexed();
+        info.contentSize = viewportIface->contentSize();
+        info.position = viewportIface->position();
+        info.viewportSize = viewportIface->viewportSize();
+
+        return info;
+    }
+
+    static jboolean populateScrollEvent(JNIEnv *env, jobject /*thiz*/, jint objectId, jobject event)
+    {
+        ScrollEventInfo info;
+        if (m_accessibilityContext) {
+            runInObjectContext(
+                    m_accessibilityContext,
+                    [objectId]() { return populateScrollEvent_helper(objectId); }, &info);
+        }
+
+        if (!info.isValid)
+            return false;
+
+        if (info.isIndexed) {
+            const int itemCount = std::max(info.contentSize.width(), info.contentSize.height());
+            env->CallVoidMethod(event, m_setItemCount, itemCount);
+            const int fromIndex = std::max(info.position.x(), info.position.y()) * itemCount;
+            env->CallVoidMethod(event, m_setFromIndex, fromIndex);
+            const int toIndex = fromIndex
+                    + std::min(info.viewportSize.width(), info.viewportSize.height() * itemCount);
+            env->CallVoidMethod(event, m_setToIndex, toIndex);
+
+        } else {
+            env->CallVoidMethod(event, m_setScrollXMethodID,
+                                (int)(info.position.x() * info.contentSize.width()));
+            env->CallVoidMethod(event, m_setMaxScrollXMethodID, (int)info.contentSize.width());
+            env->CallVoidMethod(event, m_setScrollYMethodID,
+                                (int)(info.position.y() * info.contentSize.height()));
+            env->CallVoidMethod(event, m_setMaxScrollYMethodID, (int)info.contentSize.height());
+        }
+
+        return true;
+    }
+
     static const JNINativeMethod methods[] = {
         {"setActive","(Z)V",(void*)setActive},
         {"childIdListForAccessibleObject", "(I)[I", (jintArray)childIdListForAccessibleObject},
@@ -900,6 +972,7 @@ namespace QtAndroidAccessibility
         {"screenRect", "(I)Landroid/graphics/Rect;", (jobject)screenRect},
         {"hitTest", "(FF)I", (void*)hitTest},
         {"populateNode", "(ILandroid/view/accessibility/AccessibilityNodeInfo;)Z", (void*)populateNode},
+        {"populateScrollEvent", "(ILandroid/view/accessibility/AccessibilityEvent;)Z", (void*)populateScrollEvent},
         {"clickAction", "(I)Z", (void*)clickAction},
         {"focusAction", "(I)Z", (void*)focusAction},
         {"scrollForward", "(I)Z", (void*)scrollForward},
@@ -957,6 +1030,15 @@ namespace QtAndroidAccessibility
                 m_setRangeInfoMethodID, nodeInfoClass, "setRangeInfo",
                 "(Landroid/view/accessibility/AccessibilityNodeInfo$RangeInfo;)V");
 
+        jclass eventClass = env->FindClass("android/view/accessibility/AccessibilityEvent");
+        GET_AND_CHECK_STATIC_METHOD(m_setMaxScrollXMethodID, eventClass, "setMaxScrollX", "(I)V");
+        GET_AND_CHECK_STATIC_METHOD(m_setScrollXMethodID, eventClass, "setScrollX", "(I)V");
+        GET_AND_CHECK_STATIC_METHOD(m_setMaxScrollYMethodID, eventClass, "setMaxScrollY", "(I)V");
+        GET_AND_CHECK_STATIC_METHOD(m_setScrollYMethodID, eventClass, "setScrollY", "(I)V");
+        GET_AND_CHECK_STATIC_METHOD(m_setItemCount, eventClass, "setItemCount", "(I)V");
+        GET_AND_CHECK_STATIC_METHOD(m_setFromIndex, eventClass, "setFromIndex", "(I)V");
+        GET_AND_CHECK_STATIC_METHOD(m_setToIndex, eventClass, "setToIndex", "(I)V");
+
         jclass rangeInfoClass =
                 env->FindClass("android/view/accessibility/AccessibilityNodeInfo$RangeInfo");
         CHECK_AND_INIT_STATIC_FIELD(int, RANGE_TYPE_INT, rangeInfoClass, "RANGE_TYPE_INT");
